extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/WorkerSketchFetcher.java

```diff
@@ -122,22 +128,31 @@ private CompletableFuture<Either<Long, ClusterByPartitions>> inMemoryFullSketchM
                 stageDefinition.getId().getQueryId(),
                 stageDefinition.getStageNumber()
             );
-        partitionFuture.whenComplete((result, exception) -> snapshotFuture.cancel(true));
+        partitionFuture.whenComplete((result, exception) -> {
+          if (exception != null || (result != null && result.isError())) {
+            snapshotFuture.cancel(true);
+          }
```

Comment:
Else should be an ISE ?

Reply:
Wouldn't this also be called with non null result if the future successfully returns later, which shouldn't be an ISE?

Reply:
Oops. Spot on. I somehow though the future always completes on error.